### PR TITLE
swap network namespace for a socket

### PR DESCRIPTION
**Pull request by aloktiagi**
This change enables the af_tsa kernel module to swap a sockets network namespace. 

The target network namespace to switch to will be the ipv4 transition network namespace.

The module checks whether the socket is really unconnected before making the namespace switch.

### CODE REVIEW

File: src/af_tsa3.c [hunk 1]
```diff
@@ -1017,238 +44,50 @@ static int tsa_swap(struct sk_buff *skb, struct genl_info *info)
...
-	err = tsa_create_reply(fd, reply, info);
-	if (!err) {
-		fd_install(fd, newfile);
-		err = 0;
-		goto out_put_net;
+	if (sk->sk_bound_dev_if != 0) {
+		NL_SET_ERR_MSG_MOD(info->extack, "Socket is bound to a device");
+		goto out_put_release;
 	}
+	mine = sock_net(sk);
```
Comment (MyUmmaGumma) [inline src/af_tsa3.c:71]: Can we change the variable names to `my_ns` and the other to `other_ns`?

File: src/af_tsa3.c [hunk 2]
```diff
@@ -1017,238 +44,50 @@ static int tsa_swap(struct sk_buff *skb, struct genl_info *info)
...
-
-	fd = get_unused_fd_flags(flags);
-	if (fd < 0) {
-		NL_SET_ERR_MSG_MOD(info->extack, "get_unused_fd_flags failed");
-		err = fd;
-		goto out_put_file;
+	// Ensure socket is in an unconnected state and is
+	// not bound to a device to ensure switching network
+	// namespaces does not result in undesired behavior.
+	sk = sock->sk;
```
Comment (MyUmmaGumma) [inline src/af_tsa3.c:60]: Do we need to check `sk` for null?

**Review by MyUmmaGumma — APPROVED** [review]

File: src/af_tsa3.c (hunk 1 above)
Comment (aloktiagi) [reply src/af_tsa3.c]: fixed

File: src/af_tsa3.c (hunk 2 above)
Comment (aloktiagi) [reply src/af_tsa3.c:60]: added a check. it shouldn't be null at this point though, but good to check.

**Review by MyUmmaGumma — APPROVED** [review]

**Comment by tych0** [comment]
Seems like there might be some loopback issues still lurking, but hopefully this won't crash :)

**Review by tych0 — APPROVED** [review]

**Comment by aloktiagi** [comment]
> Seems like there might be some loopback issues still lurking, but hopefully this won't crash :)

in the case where the socket is connected?

**Comment by tych0** [comment]
> in the case where the socket is connected?

Well, or when it gets disconnected and then we try a `setns()` after some loopback data is transferred. it looks like in e.g. ipv4_blackhole_route() we take a copy of that pointer and stick it in the routing table. this will all be invalidated when we switch network namespaces. is all that kind of thing automatically freed when a socket disconnects? i don't really know.

**Comment by aloktiagi** [comment]
> > in the case where the socket is connected?
> 
> Well, or when it gets disconnected and then we try a `setns()` after some loopback data is transferred. it looks like in e.g. ipv4_blackhole_route() we take a copy of that pointer and stick it in the routing table. this will all be invalidated when we switch network namespaces. is all that kind of thing automatically freed when a socket disconnects? i don't really know.

sure. I'll investigate that piece, though we wont intercept loopback traffic for TSA but something to make sure of.

**Pull request merged**